### PR TITLE
Preallocate buffers for Falcon and Mixtral CCLs

### PR DESCRIPTION
**Pull request by yugi957**
### Ticket
#21726 

### Problem description
Need to get rid of all legacy usage in Falcon 40b and Mixtral on T3K and use Fabric instead, this PR specifically creates all the persistent buffers necessary for those CCLs on model compilation rather than during runtime. Creating intermediate/output tensors during runtime creates race conditions within CCL ops, thus the need to preallocate PBs.

### What's changed
Extracted shapes and mem_configs of PBs going into Fabric ccls and create them on compilation.

(This should be merged after the commits authored by Grayson are merged)

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by SeanNijjar** [comment]
Is the CCL team planning on eventually providing caller-side equivalent interfaces as legacy CCLs (i.e. user doesn't have to manage explicit buffering or global semaphore creation?)

**Comment by ntarafdar** [comment]
> Is the CCL team planning on eventually providing caller-side equivalent interfaces as legacy CCLs (i.e. user doesn't have to manage explicit buffering or global semaphore creation?)

Ya this is something on our roadmap. the creation of persistent buffers and semaphores is a mess rn.

File: models/demos/t3000/falcon40b/demo/demo.py [hunk 1]
```diff
@@ -216,6 +216,7 @@ def run_falcon_demo_kv(
     prefill_ids, num_users, num_input_tokens = preprocess_and_validate_inputs(
         input_prompts, tokenizer, max_seq_len, perf_mode
     )
+    actual_seq_len = prefill_ids.shape[1]
```
Comment (grechsteinerTT) [inline models/demos/t3000/falcon40b/demo/demo.py:219]: Let's confirm with the model team that this change is correct and doesn't break anything else

File: models/demos/t3000/falcon40b/tt/falcon_causallm.py [hunk 2]
```diff
@@ -24,6 +24,7 @@ def __init__(
         model_config,
         tt_cache_path,
         use_global_cos_sin_cache,
+        seq_len: Optional[int] = None,
```
Comment (grechsteinerTT) [inline models/demos/t3000/falcon40b/tt/falcon_causallm.py:27]: Let's confirm with the model team that this change is necessary

File: models/demos/t3000/falcon40b/tt/falcon_ccl.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    def get_and_cycle_rs_semaphore_handles(self):
+        current_idx = self.rs_semaphores_idx
+        self.rs_semaphores_idx = (self.rs_semaphores_idx + 1) % 2
+        return self.rs_semaphore_handles[current_idx]
+
+    def create_persistent_buffer(self, shape, mem_config, dtype, distributed=False):
+        if distributed:
+            shape[3] *= self.mesh_device.get_num_devices()
+            cluster_shape = list(self.mesh_device.shape)
+            mesh_mapper = ttnn.ShardTensor2dMesh(self.mesh_device, dims=(None, 3), mesh_shape=cluster_shape)
```
Comment (grechsteinerTT) [inline models/demos/t3000/falcon40b/tt/falcon_ccl.py]: Why is `ShardTensor2dMesh(...)` required here? I didn't have to use it for `tt_transformers` (although I could be doing something wrong over there)

File: models/demos/t3000/falcon40b/tt/falcon_ccl.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        return self.ag_semaphore_handles[current_idx]
+
+    def get_and_cycle_rs_semaphore_handles(self):
+        current_idx = self.rs_semaphores_idx
+        self.rs_semaphores_idx = (self.rs_semaphores_idx + 1) % 2
+        return self.rs_semaphore_handles[current_idx]
+
+    def create_persistent_buffer(self, shape, mem_config, dtype, distributed=False):
+        if distributed:
+            shape[3] *= self.mesh_device.get_num_devices()
```
Comment (grechsteinerTT) [inline models/demos/t3000/falcon40b/tt/falcon_ccl.py]: Can we just update the shape defined in `create_persistent_buffers()` instead of doing it here?

File: models/demos/t3000/falcon40b/tt/falcon_ccl.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+        self.ag_output_pbs["MODEL_FWD_PREFILL_AG"] = self.create_persistent_buffer(
+            shape=[1, 1, self.seq_len, 8192],
+            mem_config=self.model_config["DEFAULT_MEMCFG"],
+            dtype=self.model_config["BFP8_DTYPE"],
+        )
+
+        # decode
+        shard_shape = [32, 256]
+        mem_config = ttnn.MemoryConfig(
```
Comment (grechsteinerTT) [inline models/demos/t3000/falcon40b/tt/falcon_ccl.py:125]: Maybe these `mem_configs` belong in `model_config` as well? (since all the previous `mem_configs` seem to be defined in `model_config`

File: models/demos/t3000/falcon40b/tt/falcon_ccl.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        return ttnn.from_torch(
+            torch.zeros(shape),
+            device=self.mesh_device,
+            layout=ttnn.TILE_LAYOUT,
+            dtype=dtype,
+            memory_config=mem_config,
+            mesh_mapper=mesh_mapper,
+        )
+
+    def create_persistent_buffers(self):
```
Comment (grechsteinerTT) [inline models/demos/t3000/falcon40b/tt/falcon_ccl.py:84]: minor nit: maybe allocate all the AG buffers first and then all the RS buffers (instead of having them intertwined)

File: models/demos/t3000/mixtral8x7b/tt/mixtral_ccl.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            }
+        )
+
+        # prefill
+        self.ag_output_pbs["MOE_FWD_PREFILL_AG"] = self.create_persistent_buffer(
+            shape=[1, 8, 256, 4096],
+            mem_config=ttnn.DRAM_MEMORY_CONFIG,
+            dtype=ttnn.DataType.BFLOAT8_B,
+        )
+        self.ag_output_pbs[("ATTN_FWD_PREFILL_AG", tuple([1, 8, 256, 4096]))] = self.create_persistent_buffer(
```
Comment (grechsteinerTT) [inline models/demos/t3000/mixtral8x7b/tt/mixtral_ccl.py:97]: This seems kinda hacky having a conditional second variable in the key

File: models/demos/t3000/mixtral8x7b/tt/mixtral_moe.py [hunk 8]
```diff
@@ -131,15 +132,34 @@ def forward(self, inputs, mode="decode"):
...
+                subdevice_id=self.tt_ccl.worker_sub_device_id,
+            )
+            ttnn.synchronize_device(self.tt_ccl.mesh_device, sub_device_ids=[self.tt_ccl.worker_sub_device_id])
             results_11BH.deallocate(True)
             # Sum reduction
             output_11BH_reduced = ttnn.experimental.fast_reduce_nc(
                 output_11BH_gathered, dims=[1], output=None, compute_kernel_config=None
             )
-            output_11BH_gathered.deallocate(True)
+            # output_11BH_gathered.deallocate(True)
```
Comment (grechsteinerTT) [inline models/demos/t3000/mixtral8x7b/tt/mixtral_moe.py:153]: Should be able to just delete this

File: models/demos/t3000/mixtral8x7b/tt/mixtral_ccl.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                        self.mesh_device.compute_with_storage_grid_size().y - 1,
+                    ),
+                )
+            }
+        )
+
+        self.ag_semaphores_idx = 0
+        self.ag_semaphore_handles = [[], []]
+
+        self.rs_semaphores_idx = 0
```
Comment (grechsteinerTT) [inline models/demos/t3000/mixtral8x7b/tt/mixtral_ccl.py:34]: Should be able to remove all the RS logic from this version of `TT_CCL` since mixtral doesn't use RS

File: models/demos/t3000/falcon40b/tt/falcon_ccl.py (hunk 4 above)
Comment (grechsteinerTT) [reply models/demos/t3000/falcon40b/tt/falcon_ccl.py]: I believe this is also prone to bugs, since it looks like it assumes the operation is an AG on dim 3

Comment (yugi957) [reply models/demos/t3000/falcon40b/tt/falcon_ccl.py]: Yup, as of right now I don't use distributed tensors, I think earlier I was testing using them before RS, but I can take out that flag entirely for now

File: models/demos/t3000/falcon40b/tt/falcon_ccl.py (hunk 3 above)
Comment (yugi957) [reply models/demos/t3000/falcon40b/tt/falcon_ccl.py]: Not using it anymore, previously I tried using it for output buffer for RS, but can remove this now.

File: models/demos/t3000/falcon7b/demo_t3000.py [hunk 10]
```diff
@@ -34,6 +34,7 @@
     ],
 )
 @pytest.mark.parametrize("mesh_device", (1, 2, 3, 4, 5, 6, 7, 8), indirect=True)
+@pytest.mark.parametrize("device_params", [{"fabric_config": True}], indirect=True)
```
Comment (skhorasganiTT) [inline models/demos/t3000/falcon7b/demo_t3000.py:37]: Falcon7b does not use CCLs so this shouldn't be needed
